subprojects/security/src/main/java/org/gradle/security/internal/KeyringStripper.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.security.internal;
+
+import com.google.common.collect.ImmutableList;
+import org.bouncycastle.bcpg.PublicKeyPacket;
+import org.bouncycastle.bcpg.TrustPacket;
+import org.bouncycastle.openpgp.PGPPublicKey;
+import org.bouncycastle.openpgp.PGPPublicKeyRing;
+import org.bouncycastle.openpgp.operator.KeyFingerPrintCalculator;
+
+import java.lang.reflect.Constructor;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/**
+ * A utility class to strip unnecessary information from a keyring
+ */
+public class KeyringStripper {
+
+    private static final Constructor<PGPPublicKey> SUBKEY_CONSTRUCTOR;
```

Comment:
:grey_question:  You've pushed over the previous version, so I cannot easily compare: how come this complication?

Reply:
Keyring consists of a single master key and subkeys. The first key in keychain must be a master key. The master key contains no signature packets and can be created from publicKey packet - and it worked just fine for keychains with a single (master) key. However, subkeys should be created with publicKey packet and key signatures. Unfortunately, there is only package private constructor for this in bouncy castle lib, thus the compication.